src/wp-includes/functions.php

```diff
@@ -2579,18 +2596,79 @@ function wp_unique_filename( $dir, $filename, $unique_filename_callback = null )
 			}
 
 			if ( ! empty( $files ) ) {
-				// The extension case may have changed above.
-				$new_ext = ! empty( $ext2 ) ? $ext2 : $ext;
+				$count = count( $files );
 
 				// Ensure this never goes into infinite loop
 				// as it uses pathinfo() and regex in the check, but string replacement for the changes.
-				$count = count( $files );
-				$i     = 0;
+				$i = 0;
 
 				while ( $i <= $count && _wp_check_existing_file_names( $filename, $files ) ) {
 					$new_number = (int) $number + 1;
-					$filename   = str_replace( array( "-{$number}{$new_ext}", "{$number}{$new_ext}" ), "-{$new_number}{$new_ext}", $filename );
-					$number     = $new_number;
+
+					// If $ext is uppercase it was replaced with the lowercase version after the previous loop.
+					$filename = str_replace( array( "-{$number}{$lc_ext}", "{$number}{$lc_ext}" ), "-{$new_number}{$lc_ext}", $filename );
+
+					$number = $new_number;
+					$i++;
+				}
+			}
+		}
+
+		$alt_filenames = array();
```

Comment:
I think this line, and the `$number = '';` line at the beginning of the `} else {` if not using `$unique_filename_callback` should probably be moved up to just before that `if ( $unique_filename_callback ...` so that those two variables are initialized ready for the final `apply_filters(` if using a callback.

Otherwise they could be undefined rather than their defaults of `array()` and `''` respectively.

Reply:
Good catch, patching...